src/lib.rs

```diff
@@ -172,34 +169,93 @@ impl<R: Read> Archive<R> {
+                if part.len() == 0 {
+                    continue;
+                }
+                seen_non_empty_part = true;
                 push(&mut dst, part);
             }
+            // If any part of the filename is '..', then skip over unpacking the
+            // file to prevent directory traversal security issues.  See, e.g.:
+            // CVE-2001-1267, CVE-2002-0399, CVE-2005-1918, CVE-2007-4131
+            if contains_dot_dot {
+                continue;
+            }
+            // Skip cases where only slashes were seen, because this is
+            // effectively an empty filename.
+            if !seen_non_empty_part {
+                continue;
+            }
+
             if is_directory {
-                try!(fs::create_dir_all(&dst));
+                if let Err(e) = fs::create_dir_all(&dst) {
+                    if e.kind() != ErrorKind::AlreadyExists {
+                        return Err(e);
+                    }
+                }
             } else {
-                try!(fs::create_dir_all(&dst.parent().unwrap()));
+                if let Err(e) = fs::create_dir_all(&dst.parent().unwrap()) {
+                    if e.kind() != ErrorKind::AlreadyExists {
+                        return Err(e);
+                    }
+                }
+
                 {
                     let mut dst = try!(fs::File::create(&dst));
                     try!(io::copy(&mut file, &mut dst));
                 }
-                let mut perm = try!(fs::metadata(&dst)).permissions();
-                set_perms(&mut perm, try!(file.mode()));
-                try!(fs::set_permissions(&dst, perm));
+
+                try!(set_perms(&dst, try!(file.mode())));
             }
         }
         return Ok(());
 
-        #[cfg(unix)]
-        fn set_perms(perm: &mut fs::Permissions, mode: i32) {
+        #[cfg(all(unix, feature = "nightly"))]
+        fn set_perms(dst: &PathBuf, mode: i32) -> io::Result<()> {
             use std::os::unix::prelude::*;
+            let mut perm = try!(fs::metadata(dst)).permissions();
             perm.set_mode(mode as libc::mode_t);
+            try!(fs::set_permissions(dst, perm))
+        }
+        #[cfg(all(unix, not(feature = "nightly")))]
+        fn set_perms(dst: &PathBuf, mode: i32) -> io::Result<()> {
+            use std::ffi::CString;
+            use std::os::unix::ffi::OsStringExt;
+            let dst_cstring = try!(CString::new(dst.as_os_str().to_os_string().into_vec()));
+            unsafe {
+                libc::chmod(dst_cstring.as_ptr(), mode as libc::mode_t);
+                return Ok(())
+            }
         }
         #[cfg(windows)]
-        fn set_perms(perm: &mut fs::Permissions, mode: i32) {
+        fn set_perms(dst: &PathBuf, mode: i32) -> io::Result<()> {
+            let mut perm = try!(fs::metadata(dst)).permissions();
             perm.set_readonly(mode & 0o200 != 0o200);
+            try!(fs::set_permissions(dst, perm))
```

Comment:
Currently the build here is failing on windows because this requires an unstable feature unfortunately.